src/protocols/mod.rs

```diff
@@ -8,9 +8,43 @@
 pub mod gamespy;
 /// Reference: [Server List Ping](https://wiki.vg/Server_List_Ping)
 pub mod minecraft;
+/// Reference: [node-GameDig](https://github.com/gamedig/node-gamedig/blob/master/protocols/quake1.js)
+pub mod quake;
 /// General types that are used by all protocols.
 pub mod types;
 /// Reference: [Server Query](https://developer.valvesoftware.com/wiki/Server_queries)
 pub mod valve;
-/// Reference: [node-GameDig](https://github.com/gamedig/node-gamedig/blob/master/protocols/quake1.js)
-pub mod quake;
+
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[derive(Debug, Clone, PartialEq)]
+pub enum Protocol {
+    Gamespy(gamespy::GameSpyVersion),
+    Minecraft(Option<minecraft::MinecraftVersion>),
+    Quake(quake::QuakeVersion),
+    Valve(valve::SteamApp),
+}
+
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[derive(Debug, Clone, PartialEq)]
+pub struct GenericResponse {
+    pub server_name: Option<String>,
+    pub server_description: Option<String>,
+    pub server_game: Option<String>,
+    pub server_game_version: Option<String>,
+    pub server_map: Option<String>,
+    pub players_maximum: Option<u64>,
+    pub players_online: Option<u64>,
+    pub players_bots: Option<u64>,
+    pub has_password: Option<bool>,
+    // TODO: Add players (+rules?)
+    pub inner: SpecificResponse,
+}
```

Comment:
I was looking through all the protocol's responses to see if we have any matching fields at all, (and if I'm not mistaken) `players_maximum` and `players_online` are common fields, I think we could safely remove the `Option` on those.

Also, a big table with all the implemented protocols and what fields each of them has would be interesting to see, but this is just a note.